### PR TITLE
Additional attributes for mowers

### PR DESCRIPTION
**Pull request by Jpsy**
Two attributes have been added to mowers devices.   
Both are compiled from existing attributes:
* `error code` contains NONE if there is no error and the code from `last error code` if an error is active. This gives us a data point that simply contains clear error status information.
* `status code` contains a copy of `activity` as long as there is no error and a copy of `error code` if an error is active.  This corresponds to the behavior of the old device state in the integration of Wijnand Top.

Examples:

**No error**  
image

**Active error**  
image

**No error, `last error code` still visible**  
image

### CODE REVIEW

**Comment by osks** [comment]
Is there really any use for last error code with this? Isn't it better to use the Home Assistant history for that?

**Comment by Jpsy** [comment]
> Is there really any use for last error code with this? Isn't it better to use the Home Assistant history for that?

`last error code` is actually what the Gardena API gives us. From this the more useful attributes are derived by this PR. There is actually no data point in the API that simply gives us an active error instead of some active-or-history thingy.

We could decide not to expose `last error code` but as this is the original API data, I would tend to keep it.

**Comment by Wictorssons** [comment]
Not sure if this is posted in a suitable place but I am wondering how the activity maps to the status in the lovelace entities card?
For example "activity OK_CUTTING" shows up as "Cleaning". Is this customizable or is this one of the things you are looking into when adapting this integration to mowers?

**Comment by Jpsy** [comment]
Not really the right place as this is a pull request with a different subject. 
The possible states are predefined for each HA integration. See https://developers.home-assistant.io/docs/core/entity/vacuum/#states
The only thing that can be changed in the custom component is the mapping of mower states to available vacuum states. But you will not get rid of "Cleaning" and it is the closest we have to "Mowing".

**Comment by Jpsy** [comment]
@osks can you give me feedback if you want to remove last_error_code?
If so I will send another commit.

**Comment by grm** [comment]
Are you sure that the activity is None only when there is an error ? And not in other case ?

We should keep the last error code at least while we make sure there is no issue with 'ATTR_ERROR_CODE' and 'ATTR_STATUS_CODE'.

**Comment by osks** [comment]
Okay, let's keep error code.

**Comment by Jpsy** [comment]
> @Jpsy : Are you sure that the activity is None only when there is an error ? And not in other case ?

Nope, definitely not. This is conclusion by observation. It is very well possible that there are conditions where this simple rule fails. E.g. hibernation mode of the mower could break that logic. We may have to fine tune over time. I will observe and take care.

**Comment by Jpsy** [comment]
Sorry to create more fuzz, guys:

I just pushed another commit to this PR which may provoke your resistance.

The `... code` postfixes in the attribute names were getting onto my nerves. So I removed them all. We have "codes" here everywhere (`activity` is actually an `activity code`). So why postfix some and others not.

Total changes:
* `last error code` -> `last error`
* `error code` -> `error`
* `status code` -> `state`

Yes, I am aware that `last error code` is the wording from the API. But in the switch device you have already shortened it to `last error`, so why not in the vacuum. I am also aware that touching this attribute name introduces a breaking change. This is sad, but we should unify now, not in some months when more users are on board.

BTW: Changing from `status code` to `state` is the result of the fact that HA does not accept an attribute name of `status`. I guess it's a reserved word. But after all "state" is semantically the better choice anyway.

image

Throw your valued opinions at me, please! :D

**Comment by grm** [comment]
I won't throw you anything :)

Having code everywhere indeed does not add a better user experience, and maybe it lessen it.

Regarding the change from **status code** to **state**, I think ithe semantic is indeed better, and so enhance user XP (even if the data flow from API to the component become less clear, but hiding this complexity is part of the goals). Also as you said, it permits to have a better uniformity regarding naming, so there again we win something.

And lastly, I agree, this changes should be done sooner than later, there will be an impact for people who have done automations, and stuff like that, but the impact will require less effort now than later.

Someone raises his hand on this PR with a NOGO ?

File: custom_components/gardena_smart_system/const.py [hunk 1]
```diff
@@ -19,3 +19,5 @@
 ATTR_OPERATING_HOURS = "operating_hours"
 ATTR_LAST_ERROR = "last_error"
 ATTR_LAST_ERROR_CODE = "last_error_code"
```
Comment (grm) [inline custom_components/gardena_smart_system/const.py:21]: You should also remove this const as it is not used anymore

Comment (Jpsy) [reply custom_components/gardena_smart_system/const.py]: Absolutely! Done.

**Comment by grm** [comment]
I merge that and make a new release !
Thx a lot

Grm

**Pull request merged**